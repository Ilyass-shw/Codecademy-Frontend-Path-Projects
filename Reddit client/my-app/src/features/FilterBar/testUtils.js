import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
import store from '../../app/store'

const TestProvider = ({
  store,
  children
}) => <Provider store={store}>{children}</Provider>

export function render(ui, { store, ...otherOpts }) {
  return rtlRender(<TestProvider store={store}>{ui}</TestProvider>, otherOpts)
}

export function makeTestStore(opts = {}) {
  const origDispatch = store.dispatch
  store.dispatch = jest.fn(origDispatch)
  return store
}

// re-export everything
export * from "@testing-library/react";

