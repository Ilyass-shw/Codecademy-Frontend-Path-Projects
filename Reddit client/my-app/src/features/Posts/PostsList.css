.posts-container {
    margin-left: 4.5rem;
	background-color: rgba(24, 24, 24, 1);
    position: absolute;
    width: 100%;
    
}
.posts-container .posts-list{
    margin-top: 1.5rem;
    margin-left: 4rem;
    
}
