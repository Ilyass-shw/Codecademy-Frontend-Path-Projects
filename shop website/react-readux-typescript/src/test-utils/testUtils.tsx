import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import store, { makeCustomTestStore } from './faksestore';
import { RenderReturn, StoreType } from './types';
import { BrowserRouter } from 'react-router-dom';

// const TestWrapper: React.FC<WrapperParams> = ({
//   MyStore,
//   WithRouter,
//   children,
// }) => {
//   if (WithRouter) {
//     return (
//       <Provider store={MyStore}>
//         <BrowserRouter>{children}</BrowserRouter>
//       </Provider>
//     );
//   } else {
//     return <Provider store={MyStore}>{children}</Provider>;
//   }
// };

// ============= ============= ============= ============= =============

export const render = (
  ui: React.ReactElement,
  MyStore: StoreType,
  WithRouter?: 'withRouter',
): RenderReturn => {
  if (WithRouter) {
    return rtlRender(
      <Provider store={MyStore}>
        <BrowserRouter>{ui}</BrowserRouter>
      </Provider>,
    );
  } else {
    return rtlRender(<Provider store={MyStore}>{ui}</Provider>);
  }
};

// ============= ============= ============= ============= =============

export const renderWithRouterOnly = (ui: React.ReactElement): RenderReturn => {
  return rtlRender(<BrowserRouter>{ui}</BrowserRouter>);
};

// ============= ============= ============= ============= =============

export const makeTestStore = (
  Name?: string,
  initialState?: Record<string, unknown>,
  reducers = {},
): StoreType => {
  if (Name && initialState) {
    const CustomTestStore = makeCustomTestStore(Name, initialState, reducers);
    const origDispatch = CustomTestStore.dispatch;
    CustomTestStore.dispatch = jest.fn(origDispatch);
    return CustomTestStore;
  } else {
    const origDispatch = store.dispatch as jest.Mock;
    store.dispatch = jest.fn(origDispatch);
    return store;
  }
};

// re-export everything
export * from '@testing-library/react';
