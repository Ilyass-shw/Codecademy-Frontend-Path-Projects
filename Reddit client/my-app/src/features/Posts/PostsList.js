import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import Post from "./Post";
import "./PostsList.css";
import { fetchPosts, selectAllPosts } from "./postsSlice";
import PostsSkeleton from "../PostsSkeleton/PostsSkeleton";
import { nanoid } from "nanoid";

const PostsList = () => {
	const postsStatus = useSelector((state) => state.posts.status);
	const postError = useSelector((state) => state.posts.error);
	const posts = useSelector(selectAllPosts);

	const dispatch = useDispatch();

	useEffect(() => {
		if (postsStatus === "idle") {
			dispatch(fetchPosts());
		}
	}, [postsStatus, dispatch]);

	let content;

	if (postsStatus === "idle" || postsStatus === "Loading") {
		content = Array(3)
			.fill()
			.map(() => {
				return <PostsSkeleton key={nanoid()} />;
			});
	} else if (postsStatus === "failed") {
		content = postError;
	} else {
		content = posts.map((post) => {
			return <Post key={post.id} post={post} />;
		});
	}

	return (
		<div className="posts-container">
			<div className="posts-list">{content}</div>
		</div>
	);
};

export default PostsList;
