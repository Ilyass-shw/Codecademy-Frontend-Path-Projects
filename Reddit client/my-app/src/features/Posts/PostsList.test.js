import React from "react";
import PostsList from "./PostsList";
import { jest } from "@jest/globals";
import { render, makeTestStore, fireEvent, screen } from "../testUtils/testUtils";
import { fetchPosts } from "../posts/postsSlice";

describe("PostsSkeleton", () => {
	it("should render first time (with posts.status = idle, in store) without crashing ", () => {
		const initialState = {
			status: "idle",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);

		render(<PostsList />, { store });
	});

	it("should fetch posts when rendered ", () => {
		const initialState = {
			status: "idle",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		console.log(fetchPosts());

		render(<PostsList />, { store });

		expect(store.dispatch).toHaveBeenCalledTimes(1);
	// 	// expect(fetchPosts).toHaveBeenCalledTimes(2);
	// 	// expect(store.dispatch).toHaveBeenCalledWith(fetchPosts());
	});

	it("should display skeleton and match the snapshot when while posts are not fetched yet ", () => {
		const initialState = {
			status: "idle",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		const { container } = render(<PostsList />, { store });
		expect(container).toMatchSnapshot();
	});

	it("should display skeleton and match the snapshot when while posts fetching is not completed yet ", () => {
		const initialState = {
			status: "Loading",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		const { container } = render(<PostsList />, { store });
		expect(container).toMatchSnapshot();
	});

	it("should display error when posts fetching fails", () => {
		const initialState = {
			status: "failed",
			error: 'I have failed you master!'
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		const { container } = render(<PostsList />, { store });
		expect(container).toMatchSnapshot();
	});
});
