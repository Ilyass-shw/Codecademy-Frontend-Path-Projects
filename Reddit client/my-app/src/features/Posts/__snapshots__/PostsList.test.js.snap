// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsSkeleton should display error when posts fetching fails 1`] = `
<div>
  <div
    class="posts-container"
  >
    <div
      class="posts-list"
    >
      I have failed you master!
    </div>
  </div>
</div>
`;

exports[`PostsSkeleton should display skeleton and match the snapshot when while posts are not fetched yet  1`] = `
<div>
  <div
    class="posts-container"
  >
    <div
      class="posts-list"
    >
      <div>
        <div
          class="css-18kc0t3-SkeletonTheme"
        >
          <div
            class="post-block skeleton-post-block"
          >
            <div
              class="content-image skeleton-content-image"
            >
              <span>
                <span
                  class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                  style="width: 240px; height: 10rem;"
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              class="post-data skeleton-post-data-mobile"
            >
              <div
                class="post-data-container skeleton-post-data-container-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 180px; height: 20px;"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 140px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                class="subreddit skeleton-subreddit-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton subreddit-icon css-1eg0sfo-skeletonStyles-Skeleton"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 80px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="css-18kc0t3-SkeletonTheme"
        >
          <div
            class="post-block skeleton-post-block"
          >
            <div
              class="content-image skeleton-content-image"
            >
              <span>
                <span
                  class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                  style="width: 240px; height: 10rem;"
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              class="post-data skeleton-post-data-mobile"
            >
              <div
                class="post-data-container skeleton-post-data-container-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 180px; height: 20px;"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 140px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                class="subreddit skeleton-subreddit-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton subreddit-icon css-1eg0sfo-skeletonStyles-Skeleton"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 80px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="css-18kc0t3-SkeletonTheme"
        >
          <div
            class="post-block skeleton-post-block"
          >
            <div
              class="content-image skeleton-content-image"
            >
              <span>
                <span
                  class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                  style="width: 240px; height: 10rem;"
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              class="post-data skeleton-post-data-mobile"
            >
              <div
                class="post-data-container skeleton-post-data-container-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 180px; height: 20px;"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 140px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                class="subreddit skeleton-subreddit-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton subreddit-icon css-1eg0sfo-skeletonStyles-Skeleton"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 80px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PostsSkeleton should display skeleton and match the snapshot when while posts fetching is not completed yet  1`] = `
<div>
  <div
    class="posts-container"
  >
    <div
      class="posts-list"
    >
      <div>
        <div
          class="css-18kc0t3-SkeletonTheme"
        >
          <div
            class="post-block skeleton-post-block"
          >
            <div
              class="content-image skeleton-content-image"
            >
              <span>
                <span
                  class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                  style="width: 240px; height: 10rem;"
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              class="post-data skeleton-post-data-mobile"
            >
              <div
                class="post-data-container skeleton-post-data-container-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 180px; height: 20px;"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 140px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                class="subreddit skeleton-subreddit-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton subreddit-icon css-1eg0sfo-skeletonStyles-Skeleton"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 80px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="css-18kc0t3-SkeletonTheme"
        >
          <div
            class="post-block skeleton-post-block"
          >
            <div
              class="content-image skeleton-content-image"
            >
              <span>
                <span
                  class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                  style="width: 240px; height: 10rem;"
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              class="post-data skeleton-post-data-mobile"
            >
              <div
                class="post-data-container skeleton-post-data-container-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 180px; height: 20px;"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 140px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                class="subreddit skeleton-subreddit-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton subreddit-icon css-1eg0sfo-skeletonStyles-Skeleton"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 80px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="css-18kc0t3-SkeletonTheme"
        >
          <div
            class="post-block skeleton-post-block"
          >
            <div
              class="content-image skeleton-content-image"
            >
              <span>
                <span
                  class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                  style="width: 240px; height: 10rem;"
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              class="post-data skeleton-post-data-mobile"
            >
              <div
                class="post-data-container skeleton-post-data-container-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 180px; height: 20px;"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 140px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                class="subreddit skeleton-subreddit-mobile"
              >
                <span>
                  <span
                    class="react-loading-skeleton subreddit-icon css-1eg0sfo-skeletonStyles-Skeleton"
                  >
                    ‌
                  </span>
                </span>
                <span>
                  <span
                    class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
                    style="width: 80px; height: 13px;"
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
